### PR TITLE
platform/machine/qemu/cluster: create machine with multiple devices

### PR DESCRIPTION
**Pull request by arithx**
Adds the ability to create a machine with multiple disk drives
attached. Default functionality is unchanged as the original
`NewMachine` function has been renamed to `CreateMachine` which
takes an additional argument of type `QEMUOptions`. A wrapper
function now exists at `NewMachine` which calls `CreateMachine` with
an empty `QEMUOptions` struct.

This only implements a `Disk` option on the `QEMUOptions` struct,
which itself is a struct containing a `Size` string.

Whenever additional disks are present they will have a new `qcow2`
image created for them which is immediately cleaned up similar to
how we are doing for the primary disk currently.

### CODE REVIEW

File: platform/machine/qemu/cluster.go [hunk 1]
```diff
@@ -203,6 +220,21 @@ func (qc *Cluster) NewMachine(userdata *conf.UserData) (platform.Machine, error)
...
+		additionalDiskFile, err := setupDisk(disk.Size, "")
+		if err != nil {
+			return nil, err
+		}
+		defer additionalDiskFile.Close()
+		disks = append(disks, additionalDiskFile)
+		fdnum := i + 5 // fd 4 is the primary drive
+		fdset := i + 2 // fdset 1 is used by the primary drive
+		id := fmt.Sprintf("d%d", fdnum)
+		qmCmd = append(qmCmd, "-add-fd", fmt.Sprintf("fd=%d,set=%d", fdnum, fdset),
```
Comment (bgilbert) [inline platform/machine/qemu/cluster.go]: Code duplication here is not ideal.  Maybe create a closure which allocates the fd and fdset numbers and appends the args, then call it for both the primary and secondary disks.

File: cmd/kola/spawn.go [hunk 2]
```diff
@@ -87,7 +91,20 @@ func doSpawn(cmd *cobra.Command, args []string) error {
...
-		mach, err := cluster.NewMachine(userdata)
+		var mach platform.Machine
+		if kolaPlatform == "qemu" && spawnMachineOptions != "" {
+			b, err := ioutil.ReadFile(spawnMachineOptions)
+			if err != nil {
+				return fmt.Errorf("Could not read machine options: %v", err)
+			}
+
+			var machineOpts qemu.MachineOptions
+			json.Unmarshal(b, &machineOpts)
```
Comment (ajeddeloh) [inline cmd/kola/spawn.go]: Check for errors otherwise this will silently fail and not apply the options

File: platform/machine/qemu/cluster.go [hunk 3]
```diff
@@ -186,11 +181,23 @@ func (qc *Cluster) NewMachine(userdata *conf.UserData) (platform.Machine, error)
 			"-device", qc.virtio("9p", "fsdev=cfg,mount_tag=config-2"))
 	}
 
+	fdnum := 3
```
Comment (ajeddeloh) [inline platform/machine/qemu/cluster.go]: nit: I would move these declarations down below the declaration of addDisk to make it clear they aren't being closed on by the closure.

File: cmd/kola/spawn.go [hunk 4]
```diff
@@ -87,7 +91,23 @@ func doSpawn(cmd *cobra.Command, args []string) error {
...
+
+			var machineOpts qemu.MachineOptions
+			err = json.Unmarshal(b, &machineOpts)
+			if err != nil {
+				return fmt.Errorf("Could not unmarshal machine options: %v", err)
+			}
+
+			mach, err = cluster.(*qemu.Cluster).NewMachineWithOptions(userdata, machineOpts)
+		} else {
+			mach, err = cluster.NewMachine(userdata)
```
Comment (ajeddeloh) [inline cmd/kola/spawn.go]: If you specify bad machine options (e.g. an invalid suffix) mach and err are both set to nil and it segfaults later.

Comment (arithx) [reply cmd/kola/spawn.go]: Forgot to setup the err var outside the if scope. Pushed a change to fix this.

File: cmd/kola/spawn.go [hunk 5]
```diff
@@ -87,7 +91,24 @@ func doSpawn(cmd *cobra.Command, args []string) error {
 
 	var someMach platform.Machine
 	for i := 0; i < spawnNodeCount; i++ {
-		mach, err := cluster.NewMachine(userdata)
+		var mach platform.Machine
+		var err error
+		if kolaPlatform == "qemu" && spawnMachineOptions != "" {
+			b, err := ioutil.ReadFile(spawnMachineOptions)
```
Comment (ajeddeloh) [inline cmd/kola/spawn.go]: err is getting redefined here I think.

**Comment by ajeddeloh** [comment]
Tested via it's use in https://github.com/coreos/mantle/pull/690. LGTM

**Comment by arithx** [comment]
Nits fixed and repushed

**Pull request merged**